src/lib.rs

```diff
@@ -1,40 +1,97 @@
-extern crate byteorder;
-#[macro_use]
-extern crate enum_primitive;
-extern crate num;
+use std::any::type_name;
+use std::io::{Read, Seek};
 
-use std::fmt;
-use std::io::Result;
+use num_traits::FromPrimitive;
+use tiff::decoder::{Decoder, DecodingResult};
+use tiff::tags::Tag;
+use tiff::TiffResult;
 
-pub mod geotiff;
-mod lowlevel;
-mod reader;
+use crate::image_data::*;
 
-pub use geotiff::TIFF;
-use reader::*;
+mod image_data;
 
-/// The GeoTIFF library reads `.tiff` files.
+macro_rules! assert_type {
+    ($result: expr, $actual: ty, $expected: ty) => {
+        $result
+            .ok_or_else(|| {
+                format!(
+                    "Cannot represent {} as {}",
+                    type_name::<$actual>(),
+                    type_name::<$expected>()
+                )
+            })
+            .unwrap()
+    };
+}
+
+/// The basic GeoTIFF struct. This includes any metadata as well as the actual image data.
 ///
-/// It is primarily used within a routing application that needs to parse digital elevation models.
-/// As such, other use cases are NOT tested (for now).
-impl TIFF {
-    /// Opens a `.tiff` file at the location indicated by `filename`.
-    pub fn open(filename: &str) -> Result<Box<TIFF>> {
-        let tiff_reader = TIFFReader;
-        tiff_reader.load(filename)
-    }
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
+    pub num_samples: usize,
+    image_data: ImageData,
+}
```

Comment:
Another thought I had was whether we should just expose the tiff [`Decoder`](https://docs.rs/tiff/0.9.1/tiff/decoder/struct.Decoder.html), and have the attributes like `dimensions`, `num_channels`, etc be accessible via methods only. Thoughts on this?

Reply:
Given that `Decoder<R>` consumes an underlying stream of type `R`, I think this would prevent us from implementing write support in the future.

For instance, I have a use case where I need to create a new, empty GeoTIFF, which is then populated with data. Finally, I want to save it to a file.

Reply:
Ah yes, I've been thinking in read-only mode for now, but you're correct that we can't put in `Decoder<R>` if implementing write support later.

> For instance, I have a use case where I need to create a new, empty GeoTIFF, which is then populated with data. Finally, I want to save it to a file.

Secretly, I would love to see this implemented soon, mostly just so we can write some roundtrip read/write tests and not have to store sample GeoTIFF files in this repo!